File: packages/sdkwork-utils-php/src/BytesUtils.php
<?php

declare(strict_types=1);

namespace Sdkwork\Utils;

final class BytesUtils
{
    /** @var list<string> */
    private const UNITS = ['B', 'KB', 'MB', 'GB', 'TB', 'PB'];

    public static function formatBytes(int $bytes, int $decimals = 1): string
    {
        $normalized = max(0, $bytes);
        if ($normalized < 1024) {
            return $normalized . ' B';
        }

        $size = (float) $normalized;
        $unitIndex = 0;
        while ($size >= 1024 && $unitIndex < count(self::UNITS) - 1) {
            $size /= 1024;
            $unitIndex++;
        }

        return sprintf('%.' . $decimals . 'f %s', $size, self::UNITS[$unitIndex]);
    }
}

File: packages/sdkwork-utils-csharp/BytesUtils.cs
namespace Sdkwork.Utils;

public static class BytesUtils
{
    private static readonly string[] Units = ["B", "KB", "MB", "GB", "TB", "PB"];

    public static string FormatBytes(long bytes, int decimals = 1)
    {
        var normalized = Math.Max(0L, bytes);
        if (normalized < 1024L)
        {
            return $"{normalized} B";
        }

        var size = (double)normalized;
        var unitIndex = 0;
        while (size >= 1024.0 && unitIndex < Units.Length - 1)
        {
            size /= 1024.0;
            unitIndex++;
        }

        return $"{size.ToString($"F{decimals}", System.Globalization.CultureInfo.InvariantCulture)} {Units[unitIndex]}";
    }
}
